### PR TITLE
VMware: Improve vmware_portgroup module

### LINKED ISSUE #38255: vmware_portgroup always fails with 'NoneType' object has no attribute 'policy'
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
vmware_portgroup

##### ANSIBLE VERSION

```
ansible 2.6.0 (devel 873a9ddf8d) last updated 2018/04/03 16:23:46 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/seit31/ansible/lib/ansible
  executable location = /home/seit31/ansible/bin/ansible
  python version = 2.7.5 (default, Nov 20 2015, 02:00:19) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```

##### CONFIGURATION
ANSIBLE_PIPELINING(/etc/ansible/ansible.cfg) = False
ANSIBLE_SSH_ARGS(/etc/ansible/ansible.cfg) = -C -o ControlMaster=auto -o ControlPersist=2000s
ANSIBLE_SSH_CONTROL_PATH(/etc/ansible/ansible.cfg) = /tmp/%%h-%%p-%%r
ANSIBLE_SSH_CONTROL_PATH_DIR(/etc/ansible/ansible.cfg) = /tmp
DEFAULT_FILTER_PLUGIN_PATH(/etc/ansible/ansible.cfg) = [u'/etc/ansible/filter_plugins']
DEFAULT_FORKS(/etc/ansible/ansible.cfg) = 5
DEFAULT_LOG_PATH(/etc/ansible/ansible.cfg) = /var/log/ansible.log
DEFAULT_ROLES_PATH(/etc/ansible/ansible.cfg) = [u'/etc/ansible/roles', u'/usr/share/ansible/roles']
HOST_KEY_CHECKING(/etc/ansible/ansible.cfg) = False
PERSISTENT_CONNECT_TIMEOUT(/etc/ansible/ansible.cfg) = 60
RETRY_FILES_ENABLED(/etc/ansible/ansible.cfg) = False

##### OS / ENVIRONMENT
N/A

##### SUMMARY
vmware_portgroup always fails with 'NoneType' object has no attribute 'policy'

##### STEPS TO REPRODUCE

```yaml
- name: create portgroups and assign to vswitches
    vmware_portgroup:
      hostname: "{{ vcenter_hostname }}"
      username: "{{ vcenter_user }}"
      password: "{{ vcenter_pass }}"
      esxi_hostname: "{{ esxi_ip }}"
      portgroup_name: "{{ item.portgroup }}"
      vlan_id: 0
      switch_name: "{{ item.vswitch }}"
      network_policy:
        promiscuous_mode: "{{ item.promiscuous | default(false) }}"
      validate_certs: false
    with_items: "{{ hostvars[inventory_hostname].network }}"
```

##### ACTUAL RESULTS
```
The full traceback is:
  File "/tmp/ansible_lFxOyA/ansible_module_vmware_portgroup.py", line 483, in main
    pyv.process_state()
  File "/tmp/ansible_lFxOyA/ansible_module_vmware_portgroup.py", line 214, in process_state
    self.add_hosts_port_group()
  File "/tmp/ansible_lFxOyA/ansible_module_vmware_portgroup.py", line 271, in add_hosts_port_group
    change = self.create_host_port_group(host, self.portgroup_name, self.vlan_id, self.switch_name, self.network_policy)
  File "/tmp/ansible_lFxOyA/ansible_module_vmware_portgroup.py", line 327, in create_host_port_group
    if self.check_network_policy_diff(desired_pgs[0].spec.policy, network_policy) or \
  File "/tmp/ansible_lFxOyA/ansible_module_vmware_portgroup.py", line 252, in check_network_policy_diff
    (current_policy.nicTeaming.policy != desired_policy.nicTeaming.policy) or \

failed: [router-test] (item={u'ip': u'192.168.5.254', u'nic': u'eth5', u'vswitch': u'L2MLX', u'portgroup': u'192.168.5.0'}) => {
    "changed": false,
    "invocation": {
        "module_args": {
            "cluster_name": null,
            "esxi_hostname": "10.1.0.76",
            "hostname": "10.1.0.22",
            "hosts": [
                "10.1.0.76"
            ],
            "network_policy": {
                "forged_transmits": null,
                "mac_changes": null,
                "promiscuous_mode": false
            },
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "port": 443,
            "portgroup_name": "192.168.5.0",
            "state": "present",
            "switch_name": "L2MLX",
            "teaming_policy": {
                "inbound_policy": false,
                "load_balance_policy": "loadbalance_srcid",
                "notify_switches": true,
                "rolling_order": false
            },
            "username": "administrator@vsphere.local",
            "validate_certs": false,
            "vlan_id": 0
        }
    },
    "item": {
        "ip": "192.168.5.254",
        "nic": "eth5",
        "portgroup": "192.168.5.0",
        "vswitch": "L2MLX"
    },
    "msg": "'NoneType' object has no attribute 'policy'"
}



```

### PR DESCRIPTION
**Pull request by ckotte**
##### SUMMARY
* Support check mode
* VLAN ID isn't required anymore
* VLAN ID 0 (no tagging) is the default
* Options match values in vSphere Client and vmware_vswitch module (https://github.com/ansible/ansible/pull/47015)
* Policy override is configured properly (Fixes https://github.com/ansible/ansible/issues/38255, https://github.com/ansible/ansible/issues/41774, https://github.com/ansible/ansible/issues/42887, and https://github.com/ansible/ansible/issues/44326)

##### ISSUE TYPE
- Bugfix Pull Request
- Feature Pull Request (Check mode support, Traffic Shaping, NIC failover order)

##### COMPONENT NAME
vmware_portgroup

##### ANSIBLE VERSION
```
ansible 2.6.3
  config file = /root/ansible-vmware/ansible.cfg
  configured module search path = [u'/root/ansible-vmware/library']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Jul 13 2018, 13:06:57) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### ADDITIONAL INFORMATION
It's now possible to create or re-configure a standard switch port group with all options you can specify in the vSphere Client. Overrides are handled properly as well.
```
- name: test portgroup
  tags: network, portgroup, test
  vmware_portgroup:
    hostname: "{{ vcenter_server }}"
    username: "{{ vcsa_username }}"
    password: "{{ vcsa_password }}"
    esxi_hostname: "{{ inventory_hostname }}"
    switch_name: "vSwitch3"
    portgroup_name: "Test"
    vlan_id: 1234
    security:
      promiscuous_mode: False
      mac_changes: False
      forged_transmits: False
    traffic_shaping:
      enabled: True
      average_bandwidth: 100000
      peak_bandwidth: 100000
      burst_size: 102400
    teaming:
      load_balancing: 'loadbalance_srcid'
      network_failure_detection: link_status_only
      notify_switches: True
      failback: True
      failover_order:
        active_nics:
          - vmnic6
        standby_nics:
          - vmnic7
    state: present
  delegate_to: localhost
```

### CODE REVIEW

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_portgroup.py]: Add `network_policy` as well so that we keep backward compatibility.

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_portgroup.py:70]: Any specific reason for remove `inbound_policy` and `rolling_order` ?

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_portgroup.py:79]: Documentation missing for `network_failure_detection`

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_portgroup.py]:
```suggestion
        """Create Port Group on a given host
```

**Review by Akasurde — CHANGES_REQUESTED** [review]
Rest LGTM.

**Comment by Akasurde** [comment]
@pdellaert @dericcrago @Im0 @tchernomax @jeking3 Could you please review this ?

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_portgroup.py:70]: `inbound_policy` is a deprecated option according to the API documentation and it's also not available in the GUI.

`rolling_order` is called `failback` in the GUI and the behavior is also the opposite: `rolling_order = False` is actually `failback = True`. We should name the options as they appear in the GUI and not in the API. The user doesn't know the API.

However, I could add them again plus some code and a deprecation note that it will be removed in 2.11.

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_portgroup.py:79]: done

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_portgroup.py]: done

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_portgroup.py]: done

File: test/integration/targets/vmware_portgroup/tasks/main.yml
Comment (jeking3) [inline test/integration/targets/vmware_portgroup/tasks/main.yml:33]: The preamble from the top of the file to here is pretty much the same for every test.  If not, it's darn close.  Think about factoring this away so someone could instead do:

```
- import_tasks: common/setup_vcsim.yml
  vars:
    clusters: 2
    hosts: 4
    vms: 20
```

Since I was asked to provide some simulator tests for open PRs I will probably do this eventually.

Comment (ckotte) [reply test/integration/targets/vmware_portgroup/tasks/main.yml:33]: This should be done globally for all other modules if @Akasurde thinks this makes sense. As far as I know, he's working on implementing a test environment with "real" ESXi hosts and vCenter servers and vcsim shouldn't be necessary anymore in the feature. 

I just extended the tests or copy'n'pasted the directory in all of my PRs. Therefore, your recommendation isn't that beneficial for me and I won't change it until it's officially required and then it should be changed for all modules in a single PR... 

Those vcsim tests are more or less just a pain ** *** *** because a lot doesn't work anyway and my local tests in my environment are much more valuable than a simulator test...

Comment (Akasurde) [reply test/integration/targets/vmware_portgroup/tasks/main.yml:33]: I am OK with `import_tasks` idea. We can do refactoring as part of separate PR where every single testcase is converted to suggested idea rather than doing it here. Let me know if it make sense.

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_portgroup.py:70]: > `inbound_policy` is a deprecated option according to the API documentation and it's also not available in the GUI.
> 
> `rolling_order` is called `failback` in the GUI and the behavior is also the opposite: `rolling_order = False` is actually `failback = True`. We should name the options as they appear in the GUI and not in the API. The user doesn't know the API.
> 

OK
> However, I could add them again plus some code and a deprecation note that it will be removed in 2.11.

That would be great. Let me know once you have that change in place.

File: test/integration/targets/vmware_portgroup/tasks/main.yml
Comment (jeking3) [reply test/integration/targets/vmware_portgroup/tasks/main.yml:33]: I totally agree, I was just brainstorming on this, didn't mean to suggest it gets done here.  Sorry I wasn't clear about that.  Even a "real" environment would not be able to exercise all the error branches.  I think what's really needed in the vmware suite is unit testing with MagicMock, but that's another effort.  Thanks.

Comment (Akasurde) [reply test/integration/targets/vmware_portgroup/tasks/main.yml:33]: @jeking3 MagicMock is nice idea. I have seen that in one of the test case in your Pyvmomi patch.

Comment (ckotte) [reply test/integration/targets/vmware_portgroup/tasks/main.yml:33]: Ok. I will resolve this conversation. Please discuss it elsewhere. I'm happy to implement it if my PR didn't get approved before you defined it and if it's a standard for all modules.

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_portgroup.py:70]: I added `inbound_policy` and `rolling_order` back again. Can you check?

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (Straevaras) [inline lib/ansible/modules/cloud/vmware/vmware_portgroup.py]: Documentation examples aren't following the change of `portgroup_name` to `portgroup`.  Seeing now that `portgroup_name` is aliased, but unsure if you want to keep it consistent for documentation sake.

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (Straevaras) [inline lib/ansible/modules/cloud/vmware/vmware_portgroup.py]: Through reintroduction of `inbound_policy` and `rolling_order`, 295 and 296 breaks the module.  These lines are assuming a teaming config is defined.  I get the following error:

```
"msg": "'NoneType' object has no attribute '__getitem__'"

The full traceback is:
  File "/tmp/ansible_rDL_Y6/ansible_module_vmware_portgroup.py", line 1023, in main
    host_portgroup = VMwareHostPortGroup(module)
  File "/tmp/ansible_rDL_Y6/ansible_module_vmware_portgroup.py", line 295, in __init__
    self.teaming_inbound_policy = self.module.params['teaming']['inbound_policy']
```

**Review by Straevaras — CHANGES_REQUESTED** [review]
One comment about document, the other comment is a bug in the code that occurs when a teaming config is not defined.

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_portgroup.py]: Right. Thx for the review.

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_portgroup.py]: I need to keep those options for compatibility. Otherwise, the PR won't be accepted. Unfortunately, they cannot be removed easily. I changed the code. Now it should work. Can you try again?

Comment (Straevaras) [reply lib/ansible/modules/cloud/vmware/vmware_portgroup.py]: Looks good, passes my tests. 👍

Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_portgroup.py]: thx

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_portgroup.py:68]: Could you please make `load_balance_policy` as alias for `load_balancing` within the code ? I am concerned about backward compatibility.

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_portgroup.py]:
```suggestion
    - 'C(inbound_policy) and C(rolling_order) are deprecated and will be removed in 2.11'
```

**Review by Akasurde — CHANGES_REQUESTED** [review]
Rest LGTM. Let me know once you are done.

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_portgroup.py:68]: Sure. Looks like I forgot this option...

Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_portgroup.py:68]: Thanks.

Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_portgroup.py:68]: There's an alias defined. No need to add that option back.

File: lib/ansible/modules/cloud/vmware/vmware_portgroup.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_portgroup.py]: fixed

**Review by Akasurde — APPROVED** [review]

**Comment by Akasurde** [comment]
@pdellaert @dericcrago @Im0 @tchernomax @jeking3 Could you please review this ? Thanks.

**Comment by olfway** [comment]
I've tried to use this updated module to setup security policies on vmware 6.0

Sometimes (probably when there are no sec overrides for portgroup in vmware)
the spec.policy.security is non-true in update_host_port_group function
and portgroup is marked as changed even if no changes needed
(because I have no security configuration for this portgroup in ansible)

I've changed the code to mark portgroup as changed in case of non-true spec.policy.security object
only when there is at least one security option set in ansible config

```
--- vmware_portgroup.py.orig	2019-01-16 23:10:46.000000000 +0200
+++ vmware_portgroup.py	2019-01-16 23:58:01.000000000 +0200
@@ -547,25 +547,21 @@
                 changed = changed_security = True
                 changed_list.append("Forged transmits")
             if changed_security:
-                if self.sec_promiscuous_mode is None:
-                    host_results['sec_promiscuous_mode_previous'] = "No override"
-                else:
+                if self.sec_promiscuous_mode is not None:
                     host_results['sec_promiscuous_mode_previous'] = promiscuous_mode_previous
-                if self.sec_mac_changes is None:
-                    host_results['sec_mac_changes_previous'] = "No override"
-                else:
+                if self.sec_mac_changes is not None:
                     host_results['sec_mac_changes'] = mac_changes_previous
-                if self.sec_forged_transmits is None:
-                    host_results['sec_forged_transmits_previous'] = "No override"
-                else:
+                if self.sec_forged_transmits is not None:
                     host_results['sec_forged_transmits_previous'] = forged_transmits_previous
         else:
-            spec.policy.security = self.create_security_policy()
-            changed = True
-            changed_list.append("Security")
-            host_results['sec_promiscuous_mode_previous'] = "No override"
-            host_results['sec_mac_changes_previous'] = "No override"
-            host_results['sec_forged_transmits_previous'] = "No override"
+            security_policy = self.create_security_policy()
+            if security_policy is not None:
+                changed = True
+                changed_list.append("Security")
+                spec.policy.security = security_policy
+                host_results['sec_promiscuous_mode_previous'] = "No override"
+                host_results['sec_mac_changes_previous'] = "No override"
+                host_results['sec_forged_transmits_previous'] = "No override"

         # Check traffic shaping
         if self.ts_enabled is None:
@@ -828,15 +824,10 @@
         Create a Network Policy
         Returns: Network Policy object
         """
-        security_policy = None
         shaping_policy = None
         teaming_policy = None

-        # Only configure security policy if an option is defined
-        if not all(option is None for option in [self.sec_promiscuous_mode,
-                                                 self.sec_mac_changes,
-                                                 self.sec_forged_transmits]):
-            security_policy = self.create_security_policy()
+        security_policy = self.create_security_policy()
         if self.ts_enabled:
             shaping_policy = self.create_shaping_policy()
         teaming_policy = self.create_teaming_policy()
@@ -854,10 +845,15 @@
         Create a Security Policy
         Returns: Security Policy object
         """
-        security_policy = vim.host.NetworkPolicy.SecurityPolicy()
-        security_policy.allowPromiscuous = self.sec_promiscuous_mode
-        security_policy.macChanges = self.sec_mac_changes
-        security_policy.forgedTransmits = self.sec_forged_transmits
+        security_policy = None
+        # Only configure security policy if an option is defined
+        if not all(option is None for option in [self.sec_promiscuous_mode,
+                                                 self.sec_mac_changes,
+                                                 self.sec_forged_transmits]):
+            security_policy = vim.host.NetworkPolicy.SecurityPolicy()
+            security_policy.allowPromiscuous = self.sec_promiscuous_mode
+            security_policy.macChanges = self.sec_mac_changes
+            security_policy.forgedTransmits = self.sec_forged_transmits
         return security_policy

     def create_shaping_policy(self):
```

**Comment by ckotte** [comment]
Can you provide the task and the output?

**Comment by olfway** [comment]
My test playbook:

```
---
  - name: vSwitch
    hosts: localhost
    connection: local
    gather_facts: no
    any_errors_fatal: yes

    tasks:
      - name: portgroups test
        local_action:
          module: vmware_portgroup
          validate_certs: no
          hostname: "{{ vmware_host }}"
          username: "{{ vmware_admin }}"
          password: "{{ vmware_password }}"
          cluster_name: "{{ vmware_cluster }}"
          switch_name: "vSwitch0"
          portgroup_name: "VM Network"
        register: vmware_portgroup_results

      - name: debug
        debug:
          var: vmware_portgroup_results
...
```

I've just created "VM Network" portgroup without any security overrides

Output with module from this pull request
```
ansible-playbook vmw-vswitch-test.yml --check
PLAY [vSwitch] ************************************************************************************************************************************************************************************************************************

TASK [portgroups test] ****************************************************************************************************************************************************************************************************************
changed: [localhost -> localhost]

TASK [debug] **************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "vmware_portgroup_results": {
        "changed": true,
        "failed": false,
        "result": {
            "vmw-host-1": {
                "changed": true,
                "msg": "Security would be changed",
                "portgroup": "VM Network",
                "sec_forged_transmits": "No override",
                "sec_forged_transmits_previous": "No override",
                "sec_mac_changes": "No override",
                "sec_mac_changes_previous": "No override",
                "sec_promiscuous_mode": "No override",
                "sec_promiscuous_mode_previous": "No override",
                "traffic_shaping": "No override",
                "vlan_id": 0,
                "vswitch": "vSwitch0"
            },
            "vmw-host-2": {
                "changed": true,
                "msg": "Security would be changed",
                "portgroup": "VM Network",
                "sec_forged_transmits": "No override",
                "sec_forged_transmits_previous": "No override",
                "sec_mac_changes": "No override",
                "sec_mac_changes_previous": "No override",
                "sec_promiscuous_mode": "No override",
                "sec_promiscuous_mode_previous": "No override",
                "traffic_shaping": "No override",
                "vlan_id": 0,
                "vswitch": "vSwitch0"
            },
            "vmw-host-3": {
                "changed": true,
                "msg": "Security would be changed",
                "portgroup": "VM Network",
                "sec_forged_transmits": "No override",
                "sec_forged_transmits_previous": "No override",
                "sec_mac_changes": "No override",
                "sec_mac_changes_previous": "No override",
                "sec_promiscuous_mode": "No override",
                "sec_promiscuous_mode_previous": "No override",
                "traffic_shaping": "No override",
                "vlan_id": 0,
                "vswitch": "vSwitch0"
            }
        }
    }
}

PLAY RECAP ****************************************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0
```

Output with my changes:
```
ansible-playbook vmw-vswitch-test.yml --check

PLAY [vSwitch] ************************************************************************************************************************************************************************************************************************

TASK [portgroups test] ****************************************************************************************************************************************************************************************************************
ok: [localhost -> localhost]

TASK [debug] **************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "vmware_portgroup_results": {
        "changed": false,
        "failed": false,
        "result": {
            "vmw-host-1": {
                "changed": false,
                "msg": "Port Group already configured properly",
                "portgroup": "VM Network",
                "sec_forged_transmits": "No override",
                "sec_mac_changes": "No override",
                "sec_promiscuous_mode": "No override",
                "traffic_shaping": "No override",
                "vlan_id": 0,
                "vswitch": "vSwitch0"
            },
            "vmw-host-2": {
                "changed": false,
                "msg": "Port Group already configured properly",
                "portgroup": "VM Network",
                "sec_forged_transmits": "No override",
                "sec_mac_changes": "No override",
                "sec_promiscuous_mode": "No override",
                "traffic_shaping": "No override",
                "vlan_id": 0,
                "vswitch": "vSwitch0"
            },
            "vmw-host-3": {
                "changed": false,
                "msg": "Port Group already configured properly",
                "portgroup": "VM Network",
                "sec_forged_transmits": "No override",
                "sec_mac_changes": "No override",
                "sec_promiscuous_mode": "No override",
                "traffic_shaping": "No override",
                "vlan_id": 0,
                "vswitch": "vSwitch0"
            }
        }
    }
}

PLAY RECAP ****************************************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0
```

Actually it would be great to generate a diff output, but I guess this could be done after this request merged

Thanks for your work!

**Comment by Akasurde** [comment]
superseded by #56382

**Pull request closed without merging**
